### PR TITLE
Rework mod content installation directories and add D2k videos

### LINKED ISSUE #6117: Windows file system locking files is failing the CD installation
install-music

I get this error as though it's not able to access the local file it just copied. When I go back to the menu and into music after the error all the music is present there, it even plays properly. Not sure what this error represents.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/3410856-windows-file-system-locking-files-is-failing-the-cd-installation?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by penev92**
Until now there hasn't been need for the videos.
#8117 introduces the first use-case.

~~I didn't add all videos, because chances are it would be a while before they are needed, and currently `\OpenRA\Content\d2k\` is a mess which I hope I (or even better - someone else ;) ) will fix soon (caused by the way `ContentInstaller`, `InstallFromCDLogic` and `InstallUtils` don't care about target directory, they just dump everything in `\OpenRA\Content\<mod>\`).~~

**EDIT:**

OK this got expanded quite a bit. Here is what you can find:
- The content installer now actually cares about what file should go in what directory. (Right now D2k's Content directory contains 3 empty directories and 260 files (without the videos) dumped in the root dir. This makes it much better.) This would also be useful for RA2, seing it has a ton of files as well.
- D2k now gets videos (that was in the first push)
- The VQA player now doesn't crash on D2k videos (that was also in the first push)
- Fixes errors when installing game assets via the Install Music button. (or at least fixes some of the cases when that happens, there is still one case left, as I mentioned [here](https://github.com/OpenRA/OpenRA/issues/6242#issuecomment-100659964))

Fixes #6117.

The added check in `Format80` is to prevent `IndexOutOfRange` exceptions in `Format80.ReplicatePrevious()`. I've been told TD and RA videos still play.

I tested RA and D2k extensively both with Copy from CD and Download assets.

Notes:
- D2k downloads 1 file more than is copied from the CD - PALETTE.BIN  (I have nothing to do with this, but we should figure out whether we need that file).
- Downloading the assets will not place the tilesets in the correct (new) directory for some reason.

### CODE REVIEW

File: OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs [hunk 1]
```diff
@@ -304,7 +304,7 @@ bool LoadAsset(string filename)
 			if (!GlobalFileSystem.Exists(filename))
 				return false;
 
-			if (Path.GetExtension(filename) == ".vqa")
+			if (Path.GetExtension(filename.ToLower()) == ".vqa")
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: `ToLowerInvariant()`.

Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/AssetBrowserLogic.cs]: `ToLowerInvariant` please

File: mods/d2k/mod.yaml [hunk 2]
```diff
@@ -128,7 +129,7 @@ ContentInstaller:
...
-	CopyFilesFromCD: music/ambush.aud, music/arakatak.aud, music/atregain.aud, music/entordos.aud, music/fightpwr.aud, music/fremen.aud, music/hark_bat.aud, music/landsand.aud, music/options.aud, music/plotting.aud, music/risehark.aud, music/robotix.aud, music/score.aud, music/soldappr.aud, music/spicesct.aud, music/undercon.aud, music/waitgame.aud
+	CopyFilesFromCD: music/ambush.aud, music/arakatak.aud, music/atregain.aud, music/entordos.aud, music/fightpwr.aud, music/fremen.aud, music/hark_bat.aud, music/landsand.aud, music/options.aud, music/plotting.aud, music/risehark.aud, music/robotix.aud, music/score.aud, music/soldappr.aud, music/spicesct.aud, music/undercon.aud, music/waitgame.aud, movies/A_BR01_E.VQA, movies/A_BR02_E.VQA, movies/A_BR03_E.VQA, movies/A_BR04_E.VQA, movies/A_BR05_E.VQA, movies/A_BR06_E.VQA, movies/A_BR07_E.VQA, movies/A_BR08_E.VQA, movies/A_BR09_E.VQA, movies/A_FINL_E.VQA, movies/A_LOSE_E.VQA, movies/A_MNTG_E.VQA
```
Comment (Mailaender) [inline mods/d2k/mod.yaml]: Just install them all at once. You won't have a second chance as they are not part of the TestFiles and we don't have a download/install or even update logic yet.

**Comment by penev92** [comment]
Some more commits have been added. You can test/review either by commit or as a whole. Enjoy ;)

**Comment by penev92** [comment]
Rebased and added the rest of the videos.

**Comment by obrakmann** [comment]
Code looks good. I couldn't test the d2k installation as I don't have any media, but the ra installation works as before. :+1:

File: mods/d2k/mod.yaml [hunk 3]
```diff
@@ -128,8 +130,14 @@ ContentInstaller:
...
+	CopyFilesFromCD:
+		Movies: movies/A_BR01_E.VQA, movies/A_BR02_E.VQA, movies/A_BR03_E.VQA, movies/A_BR04_E.VQA, movies/A_BR05_E.VQA, movies/A_BR06_E.VQA, movies/A_BR07_E.VQA, movies/A_BR08_E.VQA, movies/A_BR09_E.VQA, movies/A_FINL_E.VQA, movies/A_LOSE_E.VQA, movies/A_MNTG_E.VQA, movies/H_BR01_E.VQA, movies/H_BR02_E.VQA, movies/H_BR03_E.VQA, movies/H_BR04_E.VQA, movies/H_BR05_E.VQA, movies/H_BR06_E.VQA, movies/H_BR07_E.VQA, movies/H_BR08_E.VQA, movies/H_BR09_E.VQA, movies/H_FINL_E.VQA, movies/H_LOSE_E.VQA, movies/H_MNTG_E.VQA, movies/O_BR01_E.VQA, movies/O_BR02_E.VQA, movies/O_BR03_E.VQA, movies/O_BR04_E.VQA, movies/O_BR05_E.VQA, movies/O_BR06_E.VQA, movies/O_BR07_E.VQA, movies/O_BR08_E.VQA, movies/O_BR09_E.VQA, movies/O_FINL_E.VQA, movies/O_LOSE_E.VQA, movies/O_MNTG_E.VQA, movies/G_INT1_E.VQA, movies/G_INT2_E.VQA, movies/G_MAPS_E.VQA, movies/G_PLN2_E.VQA, movies/G_PLNT_E.VQA, movies/T_TITL_E.VQA
```
Comment (Mailaender) [inline mods/d2k/mod.yaml]: This does not work on case-sensitive file systems:

image

image

**Comment by penev92** [comment]
Should be OK now.

File: OpenRA.Mods.Common/InstallUtils.cs [hunk 4]
```diff
@@ -66,41 +90,66 @@ public static bool ExtractFromPackage(string srcPath, string package, string ann
...
 			{
-				var fromPath = Path.Combine(srcPath, file);
-				if (!File.Exists(fromPath))
+				var targetDir = folder.Key;
+
+				foreach (var file in folder.Value)
 				{
-					onError("Cannot find " + file);
-					return false;
+					var fileName = file.ToLowerInvariant();
```
Comment (Mailaender) [inline OpenRA.Mods.Common/InstallUtils.cs]: You can't assume all file names are lower case. This will break on every Unix system.

Comment (penev92) [inline OpenRA.Mods.Common/InstallUtils.cs]: OK what exactly is it you want me to do?

Comment (Mailaender) [inline OpenRA.Mods.Common/InstallUtils.cs]: See https://github.com/OpenRA/OpenRA/pull/8119#discussion_r31877227. Just change the line to what I showed you:

```
Movies: movies/a_br01_e.vqa [...]
```

I think you are over-complicating the matter.

**Comment by penev92** [comment]
Changed file names in `mod.yaml` to lowercase.
Turns out relying on the OS to behave properly for names read from a text file is not a good idea...
Also as far as I understand, the way you mount the disk on certain OS-es can affect the case of the file names (gg, really).

**Comment by Mailaender** [comment]
:white_check_mark: Works for me. :tm: 

I don't know if the different mount options are problem in practice and do not see that as a blocker.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/67f36b4c558de9ca9b58b27360527b77e11a1e9d...a46cec68e8f7ef733975b2461cdd223a3a10bd8d)

**Comment by Mailaender** [comment]
Thanks! This added many needed improvements.

**Pull request merged**